File: docs/experiments/grok-app/native-v2/android/app/src/main/java/tv/whisco/app/HomeTime.kt
package tv.whisco.app

import java.util.Calendar
import java.util.TimeZone

object HomeTime {
    private val languageMap = mapOf(
        "Hindi" to ("Asia/Kolkata" to "Mumbai"),
        "Malayalam" to ("Asia/Kolkata" to "Kochi"),
        "Tamil" to ("Asia/Kolkata" to "Chennai"),
        "Telugu" to ("Asia/Kolkata" to "Hyderabad"),
        "Punjabi" to ("Asia/Kolkata" to "Amritsar"),
        "Bengali" to ("Asia/Dhaka" to "Dhaka"),
        "Urdu" to ("Asia/Karachi" to "Karachi"),
        "Arabic" to ("Africa/Cairo" to "Cairo"),
        "Filipino" to ("Asia/Manila" to "Manila"),
        "English" to ("Asia/Dubai" to "Gulf"),
        "Turkish" to ("Europe/Istanbul" to "Istanbul"),
        "Indonesian" to ("Asia/Jakarta" to "Jakarta"),
        "Nepali" to ("Asia/Kathmandu" to "Kathmandu"),
        "Sinhala" to ("Asia/Colombo" to "Colombo"),
    )

    private val countryMap = mapOf(
        "Bahrain" to ("Asia/Bahrain" to "Manama"),
        "BH" to ("Asia/Bahrain" to "Manama"),
        "UAE" to ("Asia/Dubai" to "Dubai"),
        "AE" to ("Asia/Dubai" to "Dubai"),
        "United Arab Emirates" to ("Asia/Dubai" to "Dubai"),
        "Saudi Arabia" to ("Asia/Riyadh" to "Riyadh"),
        "SA" to ("Asia/Riyadh" to "Riyadh"),
        "Kuwait" to ("Asia/Kuwait" to "Kuwait City"),
        "KW" to ("Asia/Kuwait" to "Kuwait City"),
        "Qatar" to ("Asia/Qatar" to "Doha"),
        "QA" to ("Asia/Qatar" to "Doha"),
        "Oman" to ("Asia/Muscat" to "Muscat"),
        "OM" to ("Asia/Muscat" to "Muscat"),
        "Egypt" to ("Africa/Cairo" to "Cairo"),
        "India" to ("Asia/Kolkata" to "Mumbai"),
        "Pakistan" to ("Asia/Karachi" to "Karachi"),
        "Bangladesh" to ("Asia/Dhaka" to "Dhaka"),
        "Philippines" to ("Asia/Manila" to "Manila"),
        "Turkey" to ("Europe/Istanbul" to "Istanbul"),
    )

    fun pair(language: String, country: String, clockSource: String?): Pair<String, String> {
        if (clockSource == "gulf") return countryMap[country] ?: ("Asia/Bahrain" to "Manama")
        return languageMap[language] ?: ("Asia/Dubai" to "Gulf")
    }

    fun format(tzId: String, now: Long = System.currentTimeMillis()): String {
        val cal = Calendar.getInstance(TimeZone.getTimeZone(tzId))
        cal.timeInMillis = now
        val h = cal.get(Calendar.HOUR).let { if (it == 0) 12 else it }
        val min = cal.get(Calendar.MINUTE).toString().padStart(2, '0')
        val am = cal.get(Calendar.AM_PM) == Calendar.AM
        return "$h:$min ${if (am) "AM" else "PM"}"
    }

    fun label(language: String, country: String = "BH", clockSource: String? = null, now: Long = System.currentTimeMillis()): String {
        val (tz, city) = pair(language, country, clockSource)
        return "${format(tz, now)} in $city"
    }

    fun channelLabel(ch: ChannelDto, now: Long = System.currentTimeMillis()): String {
        ch.country?.let { countryMap[it] }?.let { (tz, city) -> return "${format(tz, now)} in $city" }
        ch.language?.let { return label(it, now = now) }
        return ""
    }

    fun greeting(tzId: String, now: Long = System.currentTimeMillis()): String {
        val cal = Calendar.getInstance(TimeZone.getTimeZone(tzId))
        cal.timeInMillis = now
        return when (cal.get(Calendar.HOUR_OF_DAY)) {
            in 0..4 -> "Still up"
            in 5..11 -> "Good morning"
            in 12..16 -> "Good afternoon"
            in 17..21 -> "Good evening"
            else -> "Late night"
        }
    }

    fun countryName(code: String) = Pack.countries.firstOrNull { it.first == code }?.second ?: code
    fun countryCity(code: String) = countryMap[code]?.second ?: "Gulf"
}

File: docs/experiments/grok-app/native-v2/ios/WhiscoTV/WhiscoTV/HomeTime.swift
import Foundation

enum HomeTime {
    static let languageMap: [String: (tz: String, city: String)] = [
        "Hindi": ("Asia/Kolkata", "Mumbai"),
        "Malayalam": ("Asia/Kolkata", "Kochi"),
        "Tamil": ("Asia/Kolkata", "Chennai"),
        "Telugu": ("Asia/Kolkata", "Hyderabad"),
        "Punjabi": ("Asia/Kolkata", "Amritsar"),
        "Bengali": ("Asia/Dhaka", "Dhaka"),
        "Urdu": ("Asia/Karachi", "Karachi"),
        "Arabic": ("Africa/Cairo", "Cairo"),
        "Filipino": ("Asia/Manila", "Manila"),
        "English": ("Asia/Dubai", "Gulf"),
        "Turkish": ("Europe/Istanbul", "Istanbul"),
        "Indonesian": ("Asia/Jakarta", "Jakarta"),
        "Nepali": ("Asia/Kathmandu", "Kathmandu"),
        "Sinhala": ("Asia/Colombo", "Colombo"),
    ]

    static let countryMap: [String: (tz: String, city: String)] = [
        "Bahrain": ("Asia/Bahrain", "Manama"),
        "BH": ("Asia/Bahrain", "Manama"),
        "UAE": ("Asia/Dubai", "Dubai"),
        "AE": ("Asia/Dubai", "Dubai"),
        "United Arab Emirates": ("Asia/Dubai", "Dubai"),
        "Saudi Arabia": ("Asia/Riyadh", "Riyadh"),
        "SA": ("Asia/Riyadh", "Riyadh"),
        "Kuwait": ("Asia/Kuwait", "Kuwait City"),
        "KW": ("Asia/Kuwait", "Kuwait City"),
        "Qatar": ("Asia/Qatar", "Doha"),
        "QA": ("Asia/Qatar", "Doha"),
        "Oman": ("Asia/Muscat", "Muscat"),
        "OM": ("Asia/Muscat", "Muscat"),
        "Egypt": ("Africa/Cairo", "Cairo"),
        "India": ("Asia/Kolkata", "Mumbai"),
        "Pakistan": ("Asia/Karachi", "Karachi"),
        "Bangladesh": ("Asia/Dhaka", "Dhaka"),
        "Philippines": ("Asia/Manila", "Manila"),
        "Turkey": ("Europe/Istanbul", "Istanbul"),
        "France": ("Europe/Paris", "Paris"),
        "Jordan": ("Asia/Amman", "Amman"),
        "Palestine": ("Asia/Gaza", "Gaza"),
    ]

    static func pair(language: String, country: GccCountry, clockSource: String?) -> (tz: String, city: String) {
        if clockSource == "gulf" {
            return (country.tz, country.city)
        }
        return languageMap[language] ?? ("Asia/Dubai", "Gulf")
    }

    static func format(tz: String, now: Date = Date()) -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_GB")
        f.timeZone = TimeZone(identifier: tz) ?? .current
        f.dateFormat = "h:mm a"
        return f.string(from: now)
    }

    static func label(for language: String, country: GccCountry = .BH, clockSource: String? = nil, now: Date = Date()) -> String {
        let pair = pair(language: language, country: country, clockSource: clockSource)
        return "\(format(tz: pair.tz, now: now)) in \(pair.city)"
    }

    static func channelLabel(_ channel: Channel, now: Date = Date()) -> String {
        if let country = channel.country, let pair = countryMap[country] {
            return "\(format(tz: pair.tz, now: now)) in \(pair.city)"
        }
        if let lang = channel.language {
            return label(for: lang, now: now)
        }
        return ""
    }

    static func greeting(tz: String, now: Date = Date()) -> String {
        var cal = Calendar.current
        cal.timeZone = TimeZone(identifier: tz) ?? .current
        let hour = cal.component(.hour, from: now)
        if hour < 5 { return "Still up" }
        if hour < 12 { return "Good morning" }
        if hour < 17 { return "Good afternoon" }
        if hour < 22 { return "Good evening" }
        return "Late night"
    }
}

extension GccCountry {
    var tz: String {
        switch self {
        case .BH: return "Asia/Bahrain"
        case .AE: return "Asia/Dubai"
        case .SA: return "Asia/Riyadh"
        case .KW: return "Asia/Kuwait"
        case .QA: return "Asia/Qatar"
        case .OM: return "Asia/Muscat"
        }
    }
    var city: String {
        switch self {
        case .BH: return "Manama"
        case .AE: return "Dubai"
        case .SA: return "Riyadh"
        case .KW: return "Kuwait City"
        case .QA: return "Doha"
        case .OM: return "Muscat"
        }
    }
}
